opentelemetry-api/opentelemetry/backend.py

```diff
@@ -0,0 +1,131 @@
+used instead.
+"""
+
+import sys
+import os
+import importlib
+from typing import TypeVar, Type, Callable
+from functools import wraps
+
+from .trace import Tracer
+
+### Generic private code {{{1
+
+_T = TypeVar('_T')
+
+_DEFAULT_BACKEND_MODNAME = 'opentelemetry.sdk.internal.backend_impl'
+_UNIT_TEST_IGNORE_ENV = False
+
+_tracer = None
+
+def _get_fallback_impl(api_type: Type[_T]) -> _T:
+    # TODO: Move (most of) this to module docstring.
+    """Gets the fallback implementation for ``api_type``.
+
+    ``api_type`` must be a OpenTelemetry API type like :cls:`Tracer`.
+
+    First, the function tries to find a module that provides a `get_opentelemetry_backend_impl`
+    function (with the same signature as this function). The following modules are tried:
+
+    1. `$OPENTELEMETRY_PYTHON_BACKEND_<typename>` (e.g. `OPENTELEMETRY_PYTHON_BACKEND_TRACER`)
+    2. `$OPENTELEMETRY_PYTHON_BACKEND_DEFAULT` (e.g. `OPENTELEMETRY_PYTHON_BACKEND_TRACER`)
+    3. The OpenTelemetry SDK's tracer module.
+
+    Note that if any of the environment variables is set to an nonempty value, further steps
+    are not tried, even if the modulename set there is invalid or fails to load. The no-op API
+    implementation is returned instead.
+    """
+    backend_modname = None
+    if not _UNIT_TEST_IGNORE_ENV and not sys.flags.ignore_environment:
+        backend_modname = os.getenv('OPENTELEMETRY_PYTHON_BACKEND_' + api_type.__name__.upper())
+        if not backend_modname:
+            backend_modname = os.getenv('OPENTELEMETRY_PYTHON_BACKEND_DEFAULT')
+    if not backend_modname:
+        backend_modname = _DEFAULT_BACKEND_MODNAME
+    try:
+        backend_mod = importlib.import_module(backend_modname)
+    except (ImportError, SyntaxError):
+        # TODO Log/warn
+        return api_type()
+    try:
+        # Note: We use such a long name to avoid calling a function that is not intended for this
+        # API.
+        backend_fn: Callable[[Type[_T]], object] = getattr(backend_mod, 'get_opentelemetry_backend_impl')
+    except AttributeError:
+        # TODO Log/warn
+        return api_type()
+    result = backend_fn(api_type)
+    if result and isinstance(result, api_type):
+        # TODO: Warn if backend_fn returns api_type(): It should return None to indicate using the
+        # default.
+        return result
+    return api_type()
+
+def _selectimpl(api_type: Type[_T]) -> _T:
+    impl = _get_fallback_impl(api_type)
+    assert impl
+    globals()[api_type.__name__.lower()] = impl
+    return impl
+
+def _set_backend_object(api_type: Type[_T], impl_object: _T) -> None:
+    if impl_object is None:
+        raise ValueError('None is not allowed as a backend implementation.')
+    if not isinstance(impl_object, api_type):
+        raise ValueError('The object does not implement the required base class.')
+    globals()['_' + api_type.__name__.lower()] = impl_object
+
+### Public code (basically copy & paste for each type) {{{1
```

Comment:
The academical argument against putting the global objects alongside the API class definitions is that this makes these modules contain global mutable state. I don't know if that would be a real problem in practice. Now we have the problem that the `backend` module needs to implement each API module for which a global instance should be provided.

One idea I have that solves both problems but is probably over-engineered is the following:
1. Move the existing immutable modules to an `api` subpackage.
2. Create new modules with the name of the old ones that do a wildcard import from the api submodule and additionally provide the global instance.

That way, we have immutable API modules and conveniently accessible global tracers. The dependencies would be (with the example of tracer):
1. trace depends on api.trace and the backend/loader module
2. api.trace depends on nothing (nothing backend-related at least)
3. The loader module depends on nothing (nothing api related at least)

Reply:
As far as quarantining the global mutable state, I think it would be fine to use a singleton (like java's `OpenTelemetry`) or dedicated container modules (e.g. `opentelemetry.trace.tracer.tracer`, where the first `tracer` is the module and the second is the object), but I think we have different intuitions about how "bad" it is to put the global tracer in the same module as the API class. One problem I can imagine is that it would prevent us from creating the tracer from config when the module is first imported; we'd have to modify it after the module was loaded since the implementation would need to import it first. What do you see as other problems?

The wildcard import is interesting, but it's not clear that the extra complexity is less bad than having an ugly import path for the tracer.

Reply:
> What do you see as other problems?
I haven't been able to think of any particular ones tbh, that's why I wrote "academic". 😃 Let's try putting the `tracer` getter in the api module. The "create from config" problem you mention doesn't apply now because we would create the tracer when the getter is first called.

Reply:
Actually I have found an annoyance: See comment in unit test.